### PR TITLE
Implement proof statment parsing

### PR DESCRIPTION
**Pull request by sharnoff**

### CODE REVIEW

File: src/ast/proof.rs [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+    Require(Condition<'a>),
+
+    /// A contract indicates a promise by the function; if a certain condition is met of the
+    /// inputs, a separate condition will be met by the output.
+    ///
+    /// Conditions in contracts are separated by the 'implies' operator, `=>`. Here is an example:
+    /// ```text
+    /// # x < 4 => _ < 3
+    /// ```
+    /// Note that the return value is referred to by the identifier `_`.
```
Comment (JOT85) [inline src/ast/proof.rs:74]: Just to note, I'm not a huge fan of using `_` for the output. It's a concise way to write it but I feel like it has other connotations to people who've not read it before.

File: src/ast/proof.rs [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+    Contract {
+        /// The pre-condition
+        ///
+        /// This is optional in order to allow unconditional return values. This could be written
+        /// as the following:
+        /// ```text
+        /// # => _ >= x
```
Comment (JOT85) [inline src/ast/proof.rs:81]: Implies is the right symbol to use here however `=>` doesn't jump out when used with `<=` and `>=`.
I think it's the type of syntax which is important when reading and so should stand out.
`->` may be a sensible symbol too?
Implies is probably the more correct though. How about `==>`?
I don't have strong feelings though.

File: src/ast/proof.rs [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+        lhs: Box<Expr<'a>>,
+        op: ArithOp,
+        rhs: Box<Expr<'a>>,
+    },
+
+    Named(Ident<'a>),
+
+    // TODO: It might be the case that this shouldn't be an isize, so that we can allow values
+    // above isize::MAX in proof statments, but this should be discussed further.
+    Literal(isize),
```
Comment (JOT85) [inline src/ast/proof.rs:156]: This does need discussion. `isize` is definitely fine for now though.

**Comment by JOT85** [comment]
All the parsing looks great :)

**Comment by sharnoff** [comment]
Re: Formatting - I have `rustfmt` run on save, so I believe the change to`u8_to_str` was done automatically. Something to consider.

**Comment by JOT85** [comment]
> Re: Formatting - I have `rustfmt` run on save, so I believe the change to`u8_to_str` was done automatically. Something to consider.

It wasn't! I ran `cargo fmt` and it formatted it to that in my change. I assume it only formats the current file on save.
So now u8_to_str is formatted 'properly' by rustfmt.

File: src/ast/proof.rs (hunk 1 above)
Comment (sharnoff) [reply src/ast/proof.rs:74]: That's very fair - I figured it probably makes sense for now for the sake of simpler syntax

File: src/ast/proof.rs (hunk 2 above)
Comment (sharnoff) [reply src/ast/proof.rs:81]: That all makes sense. It should be fairly trivial to change if we come up with something better, so maybe we do the classic "let's stick the mediocre thing *for now* and pretend that we'll come back and change it"

**Comment by sharnoff** [comment]
> It wasn't! I ran `cargo fmt` and it formatted it to that in my change. I assume it only formats the current file on save.

Interesting! Just for experimentation, I tried it manually as well and it changed it back! Maybe there's a config difference?

File: src/ast/proof.rs (hunk 3 above)
Comment (sharnoff) [reply src/ast/proof.rs:156]: That sounds like a reasonable plan

**Comment by JOT85** [comment]
> > It wasn't! I ran `cargo fmt` and it formatted it to that in my change. I assume it only formats the current file on save.
> 
> Interesting! Just for experimentation, I tried it manually as well and it changed it back! Maybe there's a config difference?
> 
> Edit: I don't have any applied configuration, so if it isn't that, it could perhaps be a toolchain version difference. Other than that, I have no clue

I don't have a different config either. That's very interesting!

**Pull request merged**